File: app/src/main/java/com/alston/cuteweatherapp/ScreenSlidePagerAdapter.java
package com.alston.cuteweatherapp;


import android.content.Context;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;

import android.util.Log;
import android.widget.Toast;

import java.util.ArrayList;

import timber.log.Timber;

public class ScreenSlidePagerAdapter extends FragmentStatePagerAdapter {


    private ArrayList<Fragment> locationFrag =new ArrayList<>();
    public ArrayList<String> mLocations = new ArrayList<>();

    private final String TAG ="ScreenSlidePagerAdapter";
    private Context mCtx;
    private FragmentManager fragmentManager;

    public ScreenSlidePagerAdapter(FragmentManager fm,int behavior, Context ctx) {
        super(fm,behavior);
        mCtx = ctx;
    }

    protected Boolean checkLocAlreadyExist(String newLocation){
        // remember to store another ArryaList recording the location name
        //only for easier comparison
        if (mLocations.contains(newLocation)) {
            Log.d(TAG, "location already exsist");
            return true;
        }
        return false;
    }
    public ArrayList<Fragment> getLocationFrag() {
        return locationFrag;
    }


    /**
     * Actually it created new fragment rather than just get an itme.
     */
    @NonNull
    @Override
    public Fragment getItem(int position) {
        Log.d(TAG, "Pos" + String.valueOf(position));
        if (position >= 0 && position < locationFrag.size()) {
            return locationFrag.get(position);
        } else {
            Timber.d("%d item not exist", position);
            return null;
        }
    }

    public String getLocationName(int pos){
        return mLocations.get(pos);
    }

    public Integer getFragPosition(String locName){
        int index  = mLocations.indexOf(locName);
        if(index == -1){
            return POSITION_NONE;
        }
        else
            return index;
    }

    public Integer addFrag(Fragment frag,String newLocation) {

        if(checkLocAlreadyExist(newLocation))
            return -1;

        locationFrag.add(frag);
        Timber.d("fragment size %d", locationFrag.size());
        mLocations.add(newLocation);
        notifyDataSetChanged();
        Log.d(TAG, "A new fragment is created");
        return -1;
    }


    public boolean removeFrag(int pos){
        if (getCount()==1) {
            Toast.makeText(mCtx, "至少要有一個地點", Toast.LENGTH_LONG).show();
            return false;
        }
        locationFrag.remove(pos);   // caution: Integer is not the correct type.But the warning
                                    // msg won't be given. Nothing happens.
        Log.d(TAG, mLocations.get(pos)+" to be removed from mLoc");

        mLocations.remove(pos);
        Log.d(TAG, "left");
        for(String s: mLocations){
            Log.d(TAG, s);
        }

        notifyDataSetChanged();
        return true;
    }

    @Override public int getItemPosition (Object object) {
//        int index = locationFrag.indexOf (object);
//        if (index == -1)


            return POSITION_NONE;
//        else return index;
    }

//    @Override
//    public void destroyItem(ViewGroup container, int position, Object object) {
//        fragmentManager.beginTransaction().remove((Fragment) object).commitNowAllowingStateLoss();
//    }


    @Override
    public int getCount(){
        return locationFrag.size();
    }


}//class

File: app/src/main/java/com/alston/cuteweatherapp/ScreenSlidePagerAdapter.kt
package com.alston.cuteweatherapp

import android.content.Context
import android.util.Log
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.PagerAdapter
import timber.log.Timber
import java.util.*

class ScreenSlidePagerAdapter(fm: FragmentManager, behavior: Int, private val mCtx: Context) : FragmentStatePagerAdapter(fm!!, behavior) {
    private val locationFrag:ArrayList<Fragment> = ArrayList<Fragment>()
    var mLocations = ArrayList<String>()
    private val TAG:String = "ScreenSlidePagerAdapter"
    private val fragmentManager: FragmentManager = fm
    private var mAge:Int = 0

    protected fun checkLocAlreadyExist(newLocation:String):Boolean{
        return mLocations.contains(newLocation)
    }

    override public fun getItem(position: Int): Fragment {
        Log.d(TAG, "Pos$position")

        if (position >= 0 && position < locationFrag.size){
            return locationFrag.get(position)
        }else{
            throw Exception("Out of the index")
        }
    }

    public fun getFragPosition(locName:String):Int{
        val index:Int = mLocations.indexOf(locName)
        return if(index==-1) PagerAdapter.POSITION_NONE else index
    }

    public fun addFrag(frag:Fragment, newLocation: String):Int{

        if(checkLocAlreadyExist(newLocation))
            return -1;
        locationFrag.add(frag)
        mLocations.add(newLocation)
        notifyDataSetChanged()
        Log.d(TAG, "A new fragment is created");
        Log.d(TAG, "fragment size %d" + locationFrag.size);
        return 0
    }

    override fun getItemPosition(`object`: Any): Int {
        return POSITION_NONE;
    }

    public fun removeFrag(pos:Int):Boolean{
        if (getCount()==1) {
            Toast.makeText(mCtx, "至少要有一個地點", Toast.LENGTH_LONG).show();
            return false;
        }
        locationFrag.removeAt(pos)
        Log.d(TAG, mLocations[pos] + " to be removed from mLoc")
        mLocations.removeAt(pos);
        notifyDataSetChanged()
        return true
    }

    override public fun getCount():Int{
        return locationFrag.size;
    }

    public fun getLocationName(pos:Int):String {
        return mLocations.get(pos)
    }
}//class
